src/kaleidoscope/hardware/ATMegaKeyboard.h

```diff
@@ -42,21 +42,26 @@ struct cRGB {
 #define ROW_PIN_LIST(...)  __VA_ARGS__
 #define COL_PIN_LIST(...)  __VA_ARGS__
 
-#define ATMEGA_KEYBOARD_CONFIG(ROW_PINS_, COL_PINS_)          	         \
-  static const int8_t matrix_rows = NUM_ARGS(ROW_PINS_);                 \
-  static const int8_t matrix_columns = NUM_ARGS(COL_PINS_);              \
-  static constexpr uint8_t matrix_row_pins[matrix_rows] =  ROW_PINS_;    \
-  static constexpr uint8_t matrix_col_pins[matrix_columns] =  COL_PINS_; \
-                                                                         \
-  static uint16_t previousKeyState_[matrix_rows];                        \
-  static uint16_t keyState_[matrix_rows];                                \
-  static uint16_t masks_[matrix_rows];                                   \
+#define ATMEGA_KEYBOARD_CONFIG_EXT(ROW_PINS_, COL_PINS_, MATRIX_ROWS_)      \
```

Comment:
I feel like the contortions needed to define ATMEGA_KEYBOARD_CONFIG in terms of ATMEGA_KEYBOARD_CONFIG_EXT result in less readable, maintainable code.

I think that both macros should end up calling another macro or routine that sets the common const'ed variables. (matrix_row_pin_cnt, for example, feels really weird to have leaking into the general implementation, especially if matrix_columns gets left alone.)

Reply:
Hmm... we could perhaps separate the scanner and the "keyboard" part? So we'd have `ATMegaScanner`, with its own view of rows and cols. On this, we could build a very simple `ATMegaKeyboard` that exposes pretty much the same thing as-is. The `SomeExpanderChipScanner` would do the expander scanning bits, and the ErgoDox class would bind them together. This way, nothing would leak into the `ATMegaScanner` or `ATMegaKeyboard` classes, and ErgoDox could still reuse the scanner, and perhaps even parts of `ATMegaKeyboard` (though I'm not 100% sure about this one, but we could have another class, from which both would inherit, that implements the common stuff).

How does that sound?